src/Microsoft.Framework.Primitives/StringValues.cs

```diff
@@ -114,28 +115,64 @@ private string[] GetArrayValue()
 
         int IList<string>.IndexOf(string item)
         {
-            var index = 0;
-            foreach (var value in this)
+            return IndexOf(item);
+        }
+
+        private int IndexOf(string item)
+        {
+            if (_values != null)
             {
-                if (string.Equals(value, item, StringComparison.Ordinal))
+                string[] values = _values;
+                for (int i = 0; i < values.Length; i++)
                 {
-                    return index;
+                    if (string.Equals(values[i], item, StringComparison.Ordinal))
+                    {
+                        return i;
+                    }
                 }
-                index += 1;
+                return -1;
+            }
+
+            if (_value != null)
+            {
+                return string.Equals(_value, item, StringComparison.Ordinal) ? 0 : -1;
             }
+
             return -1;
         }
 
         bool ICollection<string>.Contains(string item)
         {
-            return ((IList<string>)this).IndexOf(item) >= 0;
+            return IndexOf(item) >= 0;
         }
 
         void ICollection<string>.CopyTo(string[] array, int arrayIndex)
         {
-            for(int i = 0; i < Count; i++)
+            CopyTo(array, arrayIndex);
+        }
+
+        private void CopyTo([NotNull] string[] array, int arrayIndex)
+        {
+            if (_values != null)
+            {
+                Array.Copy(_values, 0, array, arrayIndex, _values.Length);
+                return;
+            }
+
+            if (_value != null)
             {
-                array[arrayIndex + i] = this[i];
+                if (arrayIndex < 0)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+                }
+                if (array.Length - arrayIndex < 1)
+                {
+                    throw new ArgumentException(
```

Comment:
I agree, this looks like a more descriptive message.  Though I agree that you should pass the nameof the argument instead of hardcoding it in the string.